gpMgmt/bin/go-tools/cli/commands.go

```
@@ -0,0 +1,124 @@
+package cli
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/greenplum-db/gpdb/gp/utils"
+
+	"github.com/greenplum-db/gp-common-go-libs/gplog"
+	"github.com/greenplum-db/gpdb/gp/hub"
+	"github.com/greenplum-db/gpdb/gp/idl"
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+)
+
+var (
+	configFilePath        string
+	conf                  *hub.Config
+	defaultConfigFilePath string = "%s/gp.conf"
+
+	verbose bool
+)
+
+func RootCommand() *cobra.Command {
+	root := &cobra.Command{
+		Use: "gp",
+	}
+
+	root.PersistentFlags().StringVar(&configFilePath, "config-file", fmt.Sprintf(defaultConfigFilePath, os.Getenv("GPHOME")), `Path to gp configuration file`)
+	root.PersistentFlags().BoolVar(&verbose, "verbose", false, `Provide verbose output`)
+
+	root.AddCommand(agentCmd())
+	root.AddCommand(hubCmd())
+	root.AddCommand(installCmd())
+	root.AddCommand(startCmd())
+	root.AddCommand(statusCmd())
+	root.AddCommand(stopCmd())
+
+	return root
+}
+
+/*
+ * Various helper functions used by multiple CLI commands
+ */
+
+var ReadFile = os.ReadFile
+var Unmarshal = json.Unmarshal
+var DialContext = grpc.DialContext
+
+func ParseConfig(configFilePath string) (conf *hub.Config, err error) {
+	conf = &hub.Config{}
+	conf.Credentials = &utils.GpCredentials{}
+	contents, err := ReadFile(configFilePath)
+	if err != nil {
+		return nil, err
+	}
+	err = Unmarshal(contents, &conf)
+	if err != nil {
+		return nil, err
+	}
+	return conf, nil
+}
+
+// Performs general setup needed for most commands
+// Public, so it can be mocked out in testing
+func InitializeCommand(cmd *cobra.Command, args []string) {
+	var err error
+	conf, err = ParseConfig(configFilePath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error parsing config file: %s\n", err.Error())
```

Should we use `%w` instead of `%s`?

---

The real question is why aren't we just returning `fmt.Errorf("parsing config file %q: %w", configFilePath, err)`

Its a total code smell to write directly to `os.Stderr` unless there is quite a good reason to.

---

My only guess is that we were using the `PreRun` which doesn't take an error.

We should probably use `PreRunE` instead of `PreRun`, so that we can return an error instead of us doing the `os.exit` ourselves.

---

Thanks, @Chibin for the suggestion. Changed it to use `PreRunE`. Now it is returning the error instead of printing it onto stderr